src/services/completions.ts

```diff
@@ -244,6 +244,9 @@ namespace ts.Completions {
         }
 
         function addStringLiteralCompletionsFromType(type: Type, result: CompletionEntry[]): void {
+            if (type && type.flags & TypeFlags.TypeParameter) {
+                type = (<TypeParameter>type).constraint;

```

Comment:
you should use `checker::getApparentType` here instead. it does some more than just checking the constraint.